File: file-manipulation/CSV_to_TSV.py
#----------------------------------------------------------------------#
#                  Multi-Core CSV to TSV Converter                     #
#----------------------------------------------------------------------#

# Author: Monty Perrotti

# In this program there will be use of as many cores as possible to
# convert Comma Seperated Value files to Tab Seperated Value files.

from time import perf_counter

class csv_tsv:
    input_path = ""
    output_path = ""


    def __init__(this, in_path, out_path):
        this.input_path = in_path
        this.output_path = out_path


    def run(this):
        print("Opening input file!") 
        file_csv = open(this.input_path, "rt")

        try:    # Check to see if file exists
            print("Creating output file in text mode!") 
            file_tsv = open(this.output_path, "xt")
        except:
            print("File exists! Opening output file in text write mode!") 
            file_tsv = open(this.output_path, "wt")

        for l in file_csv:
            # print(l)  # Debug
            l = l.replace(',', '\t')
            file_tsv.write(l)

        # Clean up memory
        file_csv.close()
        file_tsv.close()

tStart = perf_counter() # Used to time this program

converter = csv_tsv("defenderParts-Full.csv", "defenderParts-full.tsv")
converter.run()
tEnd = perf_counter()
execution_time = (tEnd - tStart)
print(f'Took {execution_time} seconds to run!')

File: file-manipulation/CSV_to_TSV.cpp
//----------------------------------------------------------------------#
//                  CSV to TSV Converter C++ Edition                    #
//----------------------------------------------------------------------#

// Author: Monty Perrotti


#include <iostream>
#include <fstream>
#include <thread>

#include <chrono>   // Just for timing
class CSV_to_TSV {
    public:
        CSV_to_TSV()    // Default contructor
        {
        }

        ~CSV_to_TSV()   // Default destructor
        {
        }

        void operator()(std::string inPath, std::string outPath)
        {
            convertFiles(inPath, outPath);
        }

        /**
            Converts a CSV file to a TSV file
            @param inPath string of the path to the input file
            @param outPath string of the path to the output file
        */
        void convertFiles (std::string inPath, std::string outPath)
        {
            std::ifstream inputFile;     // input CSV file
            std::ofstream outputFile;    // output TSV file

            inputFile.open(inPath, std::ios::in);
            outputFile.open(outPath, std::ios::out);
            if (!inputFile) {
                std::cout << "No such file" << std::endl;
            }
            else {
                if (!outputFile) {
                    std::cout << "File not created!" << std::endl;
                }
                else {
                    std::cout << "File created successfully!" << std::endl;
                }

                // Do replacement character by character
                // char ch; // Temporarily holds character of the file
                
                // while ( !inputFile.eof() ) {
                // 	// inputFile >> ch; // This way does not read new lines charachters

                //     inputFile.get(ch);  // Trying to use different way to get new lines
                //     if ( ch == ',' )
                //         ch = '\t';

                //     outputFile << ch;
                // 	// std::cout << ch;
                // }

                // #1 Do replacement line by line ( 2X slower then going character by character)
                // #2 Trying a new design by going through full lines by using position
                // replacement on strings. IT WORKS! This new method is 2X faster then char by char.
                std::string line;    // Temporarily holds a file line
                while (getline(inputFile, line)){
                    // replaceAll(line, ",", "\t"); // #1
                    for ( int i = 0; i < line.length(); i++ ){
                        if ( line[i] == ',' )
                            line[i] = '\t';
                    }
                    outputFile << line;
                }

            }
            inputFile.close();
            outputFile.close();
        }


        /**
            Replaces all ocurances of toSearch to replaceStr in the data string.
            @param data string to be modified
            @param toSearch string to search for inside of data
            @param replaceStr string to replace toSearch inside of data
            @author Shubham Agrawal
        */
        void replaceAll(std::string &data, const std::string &toSearch, const std::string &replaceStr)
        {
            // Get the first occurrence
            size_t pos = data.find(toSearch);
            // Repeat till end is reached
            while( pos != std::string::npos)
            {
                // Replace this occurrence of Sub String
                data.replace(pos, toSearch.size(), replaceStr);
                // Get the next occurrence from the current position
                pos = data.find(toSearch, pos + replaceStr.size());
            }
        }

};

  
int main(int argc, char *argv[])
{
    // "defenderParts-Full.csv" "defenderParts-Full.tsv"
    CSV_to_TSV converter;
    auto t_Start = std::chrono::high_resolution_clock::now();
    converter.convertFiles(argv[1], argv[2]);
    auto t_End = std::chrono::high_resolution_clock::now();

    std::chrono::microseconds duration = std::chrono::duration_cast<std::chrono::microseconds>(t_End - t_Start);

    std::cout << "Took " << duration.count() << " microseconds to run!" << std::endl;
    return 0;   // Exit app with code 0
}
